simulation/halsim_ws_core/doc/wpilib-ws.yaml

```
@@ -514,3 +515,166 @@ components:
             ">3v3_faults":
               type: integer
               description: "Number of faults on 3.3V rail"
+
+    canmotorData:
+      type: object
+      required:
+      - type
+      - device
+      properties:
+        type:
+          type: string
+          description: "Dummy type used for SimDevice type with device=CANMotor:*"
+          const: CANMotor
+        device:
+          type: string
+          description: "Should have the form CANMotor:ControllerClassName[id]"
+        data:
+          type: object
+          description: "CANMotor Data"
+          properties:
+            <init:
+              type: boolean
+              description: "If CANMotor is initialized in the robot program"
+            "<percentOutput":
+              type: number
+              description: "Percent Output"
+              minimum: -1.0
+              maximum: 1.0
+            ">supplyCurrent":
+              type: number
+              description: "Supply current in Amps"
+              minimum: 0.0
+            ">motorCurrent":
+              type: number
+              description: "Motor current in Amps"
+              minimum: 0.0
+            ">busVoltage":
+              type: number
+              description: "Bus voltage in Volts"
+              minimum: 0.0
+            "<neutralDeadband":
+              type: number
+              description: "Neutral Deadband"
+              minimum: 0
+              maximum: 1.0
+            "<brakeMode":
+              type: boolean
+              description: "Is in brake mode"
+
+    canencoderData:
+      type: object
+      required:
+      - type
+      - device
+      properties:
+        type:
+          type: string
+          description: "Dummy type used for SimDevice type with device=CANEncoder:*"
+          const: CANEncoder
+        device:
+          type: string
+          description: "Should have the form CANEncoder:ControllerClassName[id]"
+        data:
+          type: object
+          description: "CANEncoder Data"
+          properties:
+            <init:
+              type: boolean
+              description: "If CANMotor is initialized in the robot program"
+            ">position":
+              type: number
+              description: "Position in rotations"
+            ">velocity":
+              type: number
+              description: "Velocity in rotations per second"
+
+    simdeviceData:
+      type: object
+      required:
+      - type
+      - device
+      oneOf:
+      - properties:
+          type:
+            type: string
+            const: SimDevice
+          device:
+            type: string
+            format: "[^:]+"
```

It's not written in the spec that other devices cannot have colons in the name. Might I suggest (although a bit more verbose)

```suggestion
            format: "([^:]+|(?!DutyCycle|CAN(Motor|Encoder|Gyro|Accel|AIn|DIO|DutyCycle)).*:.*)+"
```

---

Latest commit uses a slightly shorter and imo clearer regex.

---

Latest commit uses a slightly shorter and imo clearer regex.

---

I agree! Your version is clearer :D